File: java/advanced/BTree.java
package advanced;

import java.util.ArrayList;
import java.util.List;

/**
 * B-Tree of minimum degree {@code t}.
 *
 * <p>A B-Tree of degree {@code t} satisfies:
 * <ol>
 *   <li>Every node has at most {@code 2t - 1} keys.</li>
 *   <li>Every non-root node has at least {@code t - 1} keys.</li>
 *   <li>Every internal node with {@code k} keys has {@code k + 1} children.</li>
 *   <li>All leaves are at the same depth.</li>
 * </ol>
 *
 * <p>Time complexities (n = total keys, h = height ≈ log_t(n)):
 * <ul>
 *   <li>search – O(t * log_t(n)) = O(log n)</li>
 *   <li>insert – O(t * log_t(n)) = O(log n)</li>
 *   <li>delete – O(t * log_t(n)) = O(log n)</li>
 * </ul>
 *
 * <p>Space complexity: O(n).
 */
public class BTree {

    // -------------------------------------------------------------------------
    // BTreeNode inner class
    // -------------------------------------------------------------------------

    /**
     * A node in the B-Tree.
     */
    static class BTreeNode {
        int[]       keys;
        BTreeNode[] children;
        int         n;        // current number of keys
        boolean     leaf;

        BTreeNode(int t, boolean leaf) {
            this.leaf     = leaf;
            this.keys     = new int[2 * t - 1];
            this.children = new BTreeNode[2 * t];
            this.n        = 0;
        }
    }

    // -------------------------------------------------------------------------
    // Fields
    // -------------------------------------------------------------------------

    private BTreeNode root;
    private final int t; // minimum degree

    // -------------------------------------------------------------------------
    // Constructor
    // -------------------------------------------------------------------------

    /**
     * Creates an empty B-Tree with minimum degree {@code t}.
     *
     * @param t minimum degree (t >= 2); a common choice is t = 3 (order 5)
     */
    public BTree(int t) {
        if (t < 2) throw new IllegalArgumentException("Minimum degree must be >= 2");
        this.t = t;
        root = new BTreeNode(t, true);
    }

    // -------------------------------------------------------------------------
    // Search
    // -------------------------------------------------------------------------

    /**
     * Returns {@code true} if {@code key} is present in the B-Tree.
     *
     * <p>Time: O(t * log_t(n)) | Space: O(log_t(n)) recursive stack.
     *
     * @param key the key to find
     * @return {@code true} if found
     */
    public boolean search(int key) {
        return searchRec(root, key);
    }

    private boolean searchRec(BTreeNode node, int key) {
        int i = 0;
        while (i < node.n && key > node.keys[i]) i++;
        if (i < node.n && key == node.keys[i]) return true;
        if (node.leaf) return false;
        return searchRec(node.children[i], key);
    }

    // -------------------------------------------------------------------------
    // Insert
    // -------------------------------------------------------------------------

    /**
     * Inserts {@code key} into the B-Tree.  Duplicate keys are ignored.
     *
     * <p>Uses a single top-down pass (pre-emptive splits), so at most one root
     * split can occur per insertion.
     *
     * <p>Time: O(t * log_t(n)) | Space: O(log_t(n)).
     *
     * @param key the key to insert
     */
    public void insert(int key) {
        if (search(key)) return; // duplicate

        BTreeNode r = root;
        if (r.n == 2 * t - 1) {
            // Root is full – split it
            BTreeNode s = new BTreeNode(t, false);
            s.children[0] = r;
            root = s;
            splitChild(s, 0, r);
            insertNonFull(s, key);
        } else {
            insertNonFull(r, key);
        }
    }

    /**
     * Inserts {@code key} into a non-full node {@code node}.
     */
    private void insertNonFull(BTreeNode node, int key) {
        int i = node.n - 1;

        if (node.leaf) {
            // Shift keys right to make room
            while (i >= 0 && key < node.keys[i]) {
                node.keys[i + 1] = node.keys[i];
                i--;
            }
            node.keys[i + 1] = key;
            node.n++;
        } else {
            // Find child to descend into
            while (i >= 0 && key < node.keys[i]) i--;
            i++;
            if (node.children[i].n == 2 * t - 1) {
                splitChild(node, i, node.children[i]);
                if (key > node.keys[i]) i++;
            }
            insertNonFull(node.children[i], key);
        }
    }

    // -------------------------------------------------------------------------
    // splitChild
    // -------------------------------------------------------------------------

    /**
     * Splits the {@code i}-th child of {@code parent} (which must be full).
     *
     * <p>The median key of the full child is promoted to {@code parent}, and the
     * right half of the child becomes a new sibling to its right.
     *
     * <p>Time: O(t) | Space: O(t).
     *
     * @param parent the parent node (must not be full)
     * @param i      index of the child to split
     * @param child  the full child node
     */
    public void splitChild(BTreeNode parent, int i, BTreeNode child) {
        BTreeNode newNode = new BTreeNode(t, child.leaf);
        newNode.n = t - 1;

        // Copy right half of child's keys to newNode
        for (int j = 0; j < t - 1; j++) {
            newNode.keys[j] = child.keys[j + t];
        }
        // Copy right half of child's children (if not a leaf)
        if (!child.leaf) {
            for (int j = 0; j < t; j++) {
                newNode.children[j] = child.children[j + t];
            }
        }
        child.n = t - 1;

        // Shift parent's children right to insert newNode
        for (int j = parent.n; j >= i + 1; j--) {
            parent.children[j + 1] = parent.children[j];
        }
        parent.children[i + 1] = newNode;

        // Shift parent's keys right to insert median
        for (int j = parent.n - 1; j >= i; j--) {
            parent.keys[j + 1] = parent.keys[j];
        }
        parent.keys[i] = child.keys[t - 1];
        parent.n++;
    }

    // -------------------------------------------------------------------------
    // Delete
    // -------------------------------------------------------------------------

    /**
     * Deletes {@code key} from the B-Tree.  Handles all cases:
     * <ol>
     *   <li>Key is in a leaf node – remove directly.</li>
     *   <li>Key is in an internal node:<br>
     *       2a. left child has &ge;t keys – replace with predecessor.<br>
     *       2b. right child has &ge;t keys – replace with successor.<br>
     *       2c. both children have exactly t-1 keys – merge, then delete.</li>
     *   <li>Key is not in the current node – ensure child has &ge;t keys before descending
     *       (rotate from sibling or merge).</li>
     * </ol>
     *
     * <p>Time: O(t * log_t(n)) | Space: O(log_t(n)).
     *
     * @param key the key to delete
     */
    public void delete(int key) {
        if (!search(key)) return;
        deleteRec(root, key);
        // If root has 0 keys, its first child becomes the new root
        if (root.n == 0 && !root.leaf) {
            root = root.children[0];
        }
    }

    private void deleteRec(BTreeNode node, int key) {
        int i = findKeyIndex(node, key);

        if (i < node.n && node.keys[i] == key) {
            // Key found in this node
            if (node.leaf) {
                // Case 1: leaf – remove directly
                removeFromLeaf(node, i);
            } else {
                // Case 2: internal node
                if (node.children[i].n >= t) {
                    // 2a: predecessor
                    int pred = getPredecessor(node, i);
                    node.keys[i] = pred;
                    deleteRec(node.children[i], pred);
                } else if (node.children[i + 1].n >= t) {
                    // 2b: successor
                    int succ = getSuccessor(node, i);
                    node.keys[i] = succ;
                    deleteRec(node.children[i + 1], succ);
                } else {
                    // 2c: merge children[i] and children[i+1]
                    merge(node, i);
                    deleteRec(node.children[i], key);
                }
            }
        } else {
            // Key not in this node – descend
            if (node.leaf) return; // key not in tree

            boolean isLastChild = (i == node.n);
            // Ensure the child we're about to descend into has >= t keys
            if (node.children[i].n < t) {
                fill(node, i);
            }
            // After fill, merging might have shifted index
            if (isLastChild && i > node.n) {
                deleteRec(node.children[i - 1], key);
            } else {
                deleteRec(node.children[i], key);
            }
        }
    }

    /** Index of the first key >= key in node. */
    private int findKeyIndex(BTreeNode node, int key) {
        int i = 0;
        while (i < node.n && node.keys[i] < key) i++;
        return i;
    }

    private void removeFromLeaf(BTreeNode node, int i) {
        for (int j = i + 1; j < node.n; j++) node.keys[j - 1] = node.keys[j];
        node.n--;
    }

    private int getPredecessor(BTreeNode node, int i) {
        BTreeNode cur = node.children[i];
        while (!cur.leaf) cur = cur.children[cur.n];
        return cur.keys[cur.n - 1];
    }

    private int getSuccessor(BTreeNode node, int i) {
        BTreeNode cur = node.children[i + 1];
        while (!cur.leaf) cur = cur.children[0];
        return cur.keys[0];
    }

    /**
     * Merges child {@code i} and child {@code i+1} of {@code node},
     * pulling down node.keys[i] as the median.
     */
    public void merge(BTreeNode node, int i) {
        BTreeNode left  = node.children[i];
        BTreeNode right = node.children[i + 1];

        // Pull median key down into left
        left.keys[t - 1] = node.keys[i];

        // Copy right's keys into left
        for (int j = 0; j < right.n; j++) left.keys[j + t] = right.keys[j];
        // Copy right's children into left
        if (!left.leaf) {
            for (int j = 0; j <= right.n; j++) left.children[j + t] = right.children[j];
        }
        left.n = 2 * t - 1;

        // Remove median key and right child from parent
        for (int j = i + 1; j < node.n; j++) node.keys[j - 1] = node.keys[j];
        for (int j = i + 2; j <= node.n; j++) node.children[j - 1] = node.children[j];
        node.n--;
    }

    /** Ensures node.children[i] has at least t keys by borrowing or merging. */
    private void fill(BTreeNode node, int i) {
        if (i > 0 && node.children[i - 1].n >= t) {
            borrowFromPrev(node, i);
        } else if (i < node.n && node.children[i + 1].n >= t) {
            borrowFromNext(node, i);
        } else {
            if (i < node.n) merge(node, i);
            else           merge(node, i - 1);
        }
    }

    private void borrowFromPrev(BTreeNode node, int i) {
        BTreeNode child  = node.children[i];
        BTreeNode sibling = node.children[i - 1];

        // Shift child's keys right
        for (int j = child.n - 1; j >= 0; j--) child.keys[j + 1] = child.keys[j];
        if (!child.leaf) {
            for (int j = child.n; j >= 0; j--) child.children[j + 1] = child.children[j];
        }
        child.keys[0] = node.keys[i - 1];
        if (!child.leaf) child.children[0] = sibling.children[sibling.n];
        node.keys[i - 1] = sibling.keys[sibling.n - 1];
        child.n++;
        sibling.n--;
    }

    private void borrowFromNext(BTreeNode node, int i) {
        BTreeNode child   = node.children[i];
        BTreeNode sibling = node.children[i + 1];

        child.keys[child.n] = node.keys[i];
        if (!child.leaf) child.children[child.n + 1] = sibling.children[0];
        node.keys[i] = sibling.keys[0];

        for (int j = 1; j < sibling.n; j++) sibling.keys[j - 1] = sibling.keys[j];
        if (!sibling.leaf) {
            for (int j = 1; j <= sibling.n; j++) sibling.children[j - 1] = sibling.children[j];
        }
        child.n++;
        sibling.n--;
    }

    // -------------------------------------------------------------------------
    // toString – level-by-level
    // -------------------------------------------------------------------------

    /**
     * Returns a level-by-level string representation of the B-Tree.
     *
     * <p>Time: O(n) | Space: O(n).
     *
     * @return multi-line tree string
     */
    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("B-Tree (t=").append(t).append("):\n");
        printLevel(sb, root, 0);
        return sb.toString();
    }

    private void printLevel(StringBuilder sb, BTreeNode node, int level) {
        if (node == null) return;
        sb.append("  ".repeat(level));
        sb.append("[");
        for (int i = 0; i < node.n; i++) {
            sb.append(node.keys[i]);
            if (i < node.n - 1) sb.append("|");
        }
        sb.append("]").append(node.leaf ? " (leaf)" : "").append("\n");
        if (!node.leaf) {
            for (int i = 0; i <= node.n; i++) {
                printLevel(sb, node.children[i], level + 1);
            }
        }
    }

    // -------------------------------------------------------------------------
    // Main – demo
    // -------------------------------------------------------------------------

    public static void main(String[] args) {
        System.out.println("=== B-Tree Demo (t=3, max 5 keys/node) ===\n");

        BTree bt = new BTree(3);

        // Insert
        int[] keys = {10, 20, 5, 6, 12, 30, 7, 17, 3, 1, 22, 25, 28, 15, 18};
        System.out.println("Inserting: ");
        for (int k : keys) {
            bt.insert(k);
            System.out.print(k + " ");
        }
        System.out.println("\n");
        System.out.println(bt);

        // Search
        System.out.println("search(6)  : " + bt.search(6));
        System.out.println("search(99) : " + bt.search(99));

        // Delete – leaf case
        System.out.println("\nDelete 6 (leaf):");
        bt.delete(6);
        System.out.println(bt);

        // Delete – internal node case
        System.out.println("Delete 12 (internal):");
        bt.delete(12);
        System.out.println(bt);

        // Delete – key requiring merge
        System.out.println("Delete 20 (may trigger merge):");
        bt.delete(20);
        System.out.println(bt);

        // Delete all remaining
        System.out.println("Deleting remaining keys one by one...");
        for (int k : keys) {
            bt.delete(k);
        }
        System.out.println("After deleting all: search(10) = " + bt.search(10));
        System.out.println(bt);
    }
}

File: python/advanced/btree.py
"""
B-Tree (Order t)
=================
A self-balancing search tree generalizing BST to nodes with many children.
Every node has between t-1 and 2t-1 keys (except root which can have as few
as 1). All leaves are at the same depth, making disk-read operations uniform.

Parameter *t* is the **minimum degree** (Cormen et al.):
    - Every non-root node has >= t-1 keys and <= 2t-1 keys.
    - Every non-root non-leaf has >= t children and <= 2t children.
    - Root has >= 1 key (if tree non-empty).

Complexities  (n = number of keys, t = minimum degree):
    - Search:  O(log_t n)  i.e. O(log n)
    - Insert:  O(log_t n)
    - Delete:  O(log_t n)
    - Space:   O(n)

Use cases: database indexing, file systems (ext4, NTFS, HFS+).
"""

from __future__ import annotations
from typing import Any, List, Optional, Tuple


class BTreeNode:
    """
    A node in the B-Tree.

    Attributes:
        keys:     sorted list of keys stored in this node.
        children: list of child BTreeNode pointers (empty for leaves).
        leaf:     True if this node has no children.
    """

    def __init__(self, leaf: bool = True) -> None:
        self.keys: List[Any] = []
        self.children: List[BTreeNode] = []
        self.leaf: bool = leaf

    def __repr__(self) -> str:
        return f"BTreeNode(keys={self.keys}, leaf={self.leaf})"


class BTree:
    """
    B-Tree of minimum degree *t*.

    - Minimum keys per non-root node: t - 1
    - Maximum keys per node:          2t - 1
    """

    def __init__(self, t: int = 2) -> None:
        if t < 2:
            raise ValueError("Minimum degree t must be >= 2")
        self._t = t
        self._root: BTreeNode = BTreeNode(leaf=True)

    # ------------------------------------------------------------------
    # Search
    # ------------------------------------------------------------------

    def search(self, key: Any) -> bool:
        """
        Return True if *key* exists in the B-Tree.

        Time:  O(log n)
        Space: O(log n) call stack
        """
        return self._search(self._root, key)

    def _search(self, node: BTreeNode, key: Any) -> bool:
        i = 0
        while i < len(node.keys) and key > node.keys[i]:
            i += 1
        if i < len(node.keys) and key == node.keys[i]:
            return True
        if node.leaf:
            return False
        return self._search(node.children[i], key)

    # ------------------------------------------------------------------
    # Insert
    # ------------------------------------------------------------------

    def insert(self, key: Any) -> None:
        """
        Insert *key* into the B-Tree.

        If root is full, splits it first (only time tree grows in height).

        Time:  O(log n)
        Space: O(t) per level during split propagation
        """
        root = self._root
        if len(root.keys) == 2 * self._t - 1:
            # Root is full — create a new root and split the old one.
            new_root = BTreeNode(leaf=False)
            new_root.children.append(self._root)
            self._split_child(new_root, 0)
            self._root = new_root
        self._insert_non_full(self._root, key)

    def _insert_non_full(self, node: BTreeNode, key: Any) -> None:
        """Insert *key* into a node guaranteed to be non-full."""
        i = len(node.keys) - 1
        if node.leaf:
            node.keys.append(None)  # placeholder
            while i >= 0 and key < node.keys[i]:
                node.keys[i + 1] = node.keys[i]
                i -= 1
            node.keys[i + 1] = key
        else:
            while i >= 0 and key < node.keys[i]:
                i -= 1
            i += 1
            if len(node.children[i].keys) == 2 * self._t - 1:
                self._split_child(node, i)
                if key > node.keys[i]:
                    i += 1
            self._insert_non_full(node.children[i], key)

    def split_child(self, parent: BTreeNode, i: int) -> None:
        """Public wrapper around _split_child (for testing)."""
        self._split_child(parent, i)

    def _split_child(self, parent: BTreeNode, i: int) -> None:
        """
        Split the full child parent.children[i] into two nodes and
        promote its median key up into *parent*.

        Time:  O(t)
        """
        t = self._t
        full_child = parent.children[i]
        new_child = BTreeNode(leaf=full_child.leaf)

        # Median key index in full_child.
        median_idx = t - 1
        median_key = full_child.keys[median_idx]

        # Right half keys go into new_child.
        new_child.keys = full_child.keys[median_idx + 1:]
        full_child.keys = full_child.keys[:median_idx]

        # Right half children go into new_child (if not leaf).
        if not full_child.leaf:
            new_child.children = full_child.children[t:]
            full_child.children = full_child.children[:t]

        # Insert median key and new child pointer into parent.
        parent.keys.insert(i, median_key)
        parent.children.insert(i + 1, new_child)

    # ------------------------------------------------------------------
    # Delete
    # ------------------------------------------------------------------

    def delete(self, key: Any) -> None:
        """
        Delete *key* from the B-Tree (no-op if absent).

        Handles all cases:
          Case 1: key in leaf node.
          Case 2: key in internal node.
            2a: left child has >= t keys → replace with predecessor.
            2b: right child has >= t keys → replace with successor.
            2c: both children have t-1 keys → merge them.
          Case 3: key not in current node — ensure the child we descend
                  into has >= t keys (borrow from sibling or merge).

        Time:  O(log n)
        """
        self._delete(self._root, key)
        # If root is now empty and has a child, shrink the tree height.
        if len(self._root.keys) == 0 and not self._root.leaf:
            self._root = self._root.children[0]

    def _delete(self, node: BTreeNode, key: Any) -> None:
        t = self._t
        i = self._find_key_index(node, key)

        if i < len(node.keys) and node.keys[i] == key:
            # Key is in this node.
            if node.leaf:
                # Case 1: simply remove.
                node.keys.pop(i)
            else:
                self._delete_from_internal(node, i)
        else:
            # Key is NOT in this node.
            if node.leaf:
                return  # key not in tree
            # Ensure child[i] has at least t keys before descending.
            if len(node.children[i].keys) < t:
                self._fill_child(node, i)
                # After fill, the structure may have shifted.
                i = self._find_key_index(node, key)
                if i < len(node.keys) and node.keys[i] == key:
                    self._delete_from_internal(node, i)
                    return
                # Recalculate child index after potential merge/borrow.
                i = min(i, len(node.children) - 1)
            self._delete(node.children[i], key)

    def _find_key_index(self, node: BTreeNode, key: Any) -> int:
        """Return the index of the first key >= *key* in node.keys."""
        i = 0
        while i < len(node.keys) and key > node.keys[i]:
            i += 1
        return i

    def _delete_from_internal(self, node: BTreeNode, i: int) -> None:
        """Delete node.keys[i] from internal node *node*."""
        t = self._t
        key = node.keys[i]
        left_child = node.children[i]
        right_child = node.children[i + 1]

        if len(left_child.keys) >= t:
            # Case 2a: replace with predecessor (max key in left subtree).
            pred = self._get_max(left_child)
            node.keys[i] = pred
            self._delete(left_child, pred)
        elif len(right_child.keys) >= t:
            # Case 2b: replace with successor (min key in right subtree).
            succ = self._get_min(right_child)
            node.keys[i] = succ
            self._delete(right_child, succ)
        else:
            # Case 2c: merge left and right children.
            self._merge_children(node, i)
            self._delete(left_child, key)

    def _get_max(self, node: BTreeNode) -> Any:
        while not node.leaf:
            node = node.children[-1]
        return node.keys[-1]

    def _get_min(self, node: BTreeNode) -> Any:
        while not node.leaf:
            node = node.children[0]
        return node.keys[0]

    def _fill_child(self, parent: BTreeNode, i: int) -> None:
        """
        Ensure parent.children[i] has at least t keys by borrowing from a
        sibling or merging with one.
        """
        t = self._t
        has_left_sibling = i > 0
        has_right_sibling = i < len(parent.children) - 1
        left_rich = has_left_sibling and len(parent.children[i - 1].keys) >= t
        right_rich = has_right_sibling and len(parent.children[i + 1].keys) >= t

        if left_rich:
            self._borrow_from_prev(parent, i)
        elif right_rich:
            self._borrow_from_next(parent, i)
        elif has_left_sibling:
            self._merge_children(parent, i - 1)
        else:
            self._merge_children(parent, i)

    def _borrow_from_prev(self, parent: BTreeNode, i: int) -> None:
        """Borrow a key from parent.children[i-1] (left sibling)."""
        child = parent.children[i]
        left_sib = parent.children[i - 1]
        # Push parent key down to front of child.
        child.keys.insert(0, parent.keys[i - 1])
        # If not leaf, move last child of left_sib to child.
        if not child.leaf:
            child.children.insert(0, left_sib.children.pop())
        # Promote last key of left_sib to parent.
        parent.keys[i - 1] = left_sib.keys.pop()

    def _borrow_from_next(self, parent: BTreeNode, i: int) -> None:
        """Borrow a key from parent.children[i+1] (right sibling)."""
        child = parent.children[i]
        right_sib = parent.children[i + 1]
        # Push parent key down to end of child.
        child.keys.append(parent.keys[i])
        # If not leaf, move first child of right_sib to child.
        if not child.leaf:
            child.children.append(right_sib.children.pop(0))
        # Promote first key of right_sib to parent.
        parent.keys[i] = right_sib.keys.pop(0)

    def _merge_children(self, parent: BTreeNode, i: int) -> None:
        """
        Merge parent.children[i] and parent.children[i+1] together,
        pulling the separator key parent.keys[i] down into the merged node.
        """
        left = parent.children[i]
        right = parent.children[i + 1]
        sep = parent.keys.pop(i)
        parent.children.pop(i + 1)

        left.keys.append(sep)
        left.keys.extend(right.keys)
        if not left.leaf:
            left.children.extend(right.children)

    # ------------------------------------------------------------------
    # ASCII visualization
    # ------------------------------------------------------------------

    def __str__(self) -> str:
        """Print the B-Tree level by level."""
        if not self._root.keys:
            return "<empty B-Tree>"
        lines = []
        level_nodes = [self._root]
        level = 0
        while level_nodes:
            level_keys = [str(n.keys) for n in level_nodes]
            lines.append(f"Level {level}: " + "  |  ".join(level_keys))
            next_level = []
            for node in level_nodes:
                next_level.extend(node.children)
            level_nodes = next_level
            level += 1
        return "\n".join(lines)


# ----------------------------------------------------------------------
# Demo
# ----------------------------------------------------------------------

if __name__ == "__main__":
    print("=== B-Tree (t=3) ===")
    bt = BTree(t=3)
    keys = [10, 20, 5, 6, 12, 30, 7, 17, 3, 1, 15, 25]
    print("Inserting:", keys)
    for k in keys:
        bt.insert(k)
    print()
    print(bt)
    print()
    print("Search 15:", bt.search(15))
    print("Search 99:", bt.search(99))
    print()
    for k in [6, 20, 10]:
        bt.delete(k)
        print(f"After deleting {k}:")
        print(bt)
        print()
